### ISSUE TITLE
Possible error in samtools tview

### ISSUE THREAD

**Issue by rlorigro** [samtools/samtools#734]
Hi,

In the process of troubleshooting my own pileup generator I have noticed that samtools sometimes generates pileups that imply nonexistent read sequences. For example, using nanopore reads ([data](https://github.com/kishwarshafin/deePore/tree/master/data)), the following pileup is generated:

```
samtools tview -p chr3:79005 -d T /hive/users/rlorigro/data/NA12878.np.chr3.100kb.0.bam /hive/users/rlorigro/data/hg19.chr3.9mb.fa 
      79011      79021      79031       79041        79051      79061           
tgttgagattacag*gcgtgag*ccactgcgcctgg*c*taattttt***gtatttttagtag*gaacagagtttcac*c
.............. ....... ............. . ........   ............. .............. .
,,,,,,,,,,,,,,*,*,c,,,*,,,,,,**,,,,,*,*,,,,,*,,***,,,,,,,,,,c,,g,,,,,,,,,,,,****
,,,,,,,,,,,,,,*****,,,*,,,,c,,,,,,,**,t,,,,,,*,gcg,,,,,*,,,,,,,**,,,,,,,,,,,,,c,
 ,,,,,,,,,,t,,g,,,,,,ttt,,a,,,,t,***c,***,c,,,,***,,,,,,,,,,,,,*,,,,,,,,**,,,,*,
```

Looking closely at this region:

```
gcctgg*c*taatt
...... . .....
*,,,,,*,*,,,,,
,,,,,**,t,,,,,
,t,***c,***,c,
```

The last alignment implies the sequence: "GTCCCACTTT"

Whereas my own interpretation of the cigar string would imply this sequence: "GTCACACTTT":

```
GTTGAGATTACAG_GCGTGAG_CCACTGCGCCTGG_C_TAATTTTT___GTATTTTTA
GTTGAGATTACAG_G*GCGAG_CCACTG**CCTGG_C_TAATT*TT___GTATTTTTA
GTTGAGATTACAG_****GAG_CCACCGCGCCTG*_CTTAATTT*TGCGGTATT*TTA
GTTGAGATTATAGGGCGTGATTTCAATGCGTC***AC_**ACTTTT___GTATTTTTA
```

Searching the SAM file with grep yields two different reads, and the sequence suggested by samtools tview doesn't exist in the correct location for it to have been mapped here:

```
>grep "GTCCCACTTT" NA12878.np.chr3.100kb.0.sam 
210fdd9b-6701-4977-b559-6b6fc15039aa_Basecall_Alignment_template	16	chr3	59965	0 ... TTGTAAGTCCCACTTTAAGAT ...
```

In other words, the sequence from samtools doesn't actually exist in the query region

```
 >grep "GTCACACTTT" NA12878.np.chr3.100kb.0.sam 
55d0825e-4caf-4c1d-8d9b-7b6416de9874_Basecall_Alignment_template	16	chr3	79006	0 ... CAATGCGTCACACTTTTGTAT ...
```

Whereas this one does.

How is it that samtools substitutes a C for an A here? Is this something related to left alignment of indels? I am using v1.5

**Comment by jkbonfield** [comment]
It's possible this is the same long standing bug reported here: https://github.com/samtools/htslib/issues/59

I'll investigate it more.  If it is the same issue, I fear it won't be easy to fix without a substantial amount of work.

**Comment by jkbonfield** [comment]
Indeed it's the same bug.  The CIGAR for this read starts 15S13M1I7M1I10M**3D1I**1M2D...  It has D and I adjacent, which is the same pileup bug we've seen before.  Issue 59 has some fixes, but I'm not sure why it never made it in - maybe fear of breaking the ABI?  I don't recall.   I'm sure I also fixed this myself at some point but cannot find that either.

(I do also have a completely rewritten pileup that I use for Gap5, which doesn't have this bug, but it would be a mammoth htslib change to switch to it.)

**Comment by rlorigro** [comment]
OK thanks for getting back to me. I am trying to incorporate pileups into my pipeline, so a separate gui implementation of this would not work for me.

**Comment by jkbonfield** [comment]
I'm not recommending anyone to use Gap5 for such things and the pileup algorithm is very different in design.  It is callback driven and generates one call per consensus column, NOT per reference location.  This means it can handle multiple alignments including insertions with P cigar operators.  However for htslib we need to consider the existing API and work out how to implement something compatible.  There are a few bits of code claiming to do this, including some of my own, but it's been so long hanging around I have no confidence in what they did and did not fix! 

I'll investigate this issue again.  Meanwhile see https://github.com/samtools/htslib/pull/226 and https://github.com/samtools/samtools/pull/331 which may be sufficient.

**Comment by jkbonfield** [comment]
I can confirm that these two PRs fix this alignment.  They've gone grotty over the last couple years with conflicts, so I'll rebase both and make some new PRs.

**Comment by valeriuo** [comment]
Fixed by https://github.com/samtools/samtools/pull/847

**Comment by tseemann** [comment]
`minimap2` is creating these `nDnI` CIGAR entries: https://github.com/lh3/minimap2/issues/460

**Comment by jkbonfield** [comment]
How ironic! ;-)
